onnx/defs/controlflow/defs.cc

```diff
@@ -227,4 +227,189 @@ ONNX_OPERATOR_SET_SCHEMA(
+    }
+    return bst_1, ..., bst_n, b_scan_out_1, ..., b_scan_out_k;
+
+
+
+*Sample usage: Encoding RNN using a Scan*
+The following example shows how a simple RNN over an input tensor %X, with weight tensor %Wi,
+recurrence weight tensor %Ri, bias tensors %Wbi and %Rbi, and initial hidden-state %H_0 can
+be encoded as a ScanLoop. Note that the loop-body is a nested graph, and it directly computes
+%Wi, %Ri, %Wbi, and %Rbi (typically constants or initializers in the body graph). If these
+values are computed in the outer graph, they need to be passed in as extra state_variables.
+
+    graph rnn-encoding {
+      %H_0 = ... 
+      %X = ...
+      %Y_h, %Y = Scan[body = <graph rnn-cell-1>]("", %H_0, %X)
+      return %Y, %Y_h
+    }
+
+    graph rnn-cell-1 (
+      %H_tminus1[FLOAT, tensor]
+      %X_t[FLOAT, tensor]
+    ) {
+      %Wi = ...
+      %Ri = ...
+      %Wbi = ...
+      %Rbi = ...
+      %t1 = X_t * (Wi^T)
+      %t2 = H_tminus1*(Ri^T)
+      %t3 = Add(%t1, %t2)
+      %t4 = Add(%t3, %Wbi)
+      %t5 = Add(%t4, %Rbi)
+      %Ht = Tanh(%t5)
+      %Accumulate = Identity(%Ht)
+      return %Ht, %Accumulate
+    }
+
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+	Scan,
+	8,
+	OpSchema()
+	.SetDoc(scan_ver1_doc)
+    .Input(
+		0,
+		"sequence_lens",
+		"Optional tensor specifying lengths of the sequences in a batch. "
+        "If this input is not specified, all sequences are assumed to be of "
+        "the maximum sequence length (the dimension of the sequence axis of "
+        "the scan_input tensors).",
+		"I",
+		OpSchema::Optional)
+	.Input(
+		1,
+		"initial_state_and_scan_inputs",
+		"Initial values of the loop's N state variables followed by M scan_inputs",
+		"V",
+		OpSchema::Variadic)
+	.Output(
+		0,
+		"final_state_and_scan_outputs",
+		"Final values of the loop's N state variables followed by K scan_outputs",
+		"V",
+		OpSchema::Variadic)
+	.Attr(
+		"body",
+		"The graph run each iteration. It has N+M inputs: "
+		"(loop state variables..., scan_input_elts...). It has N+K outputs: "
+		"(loop state variables..., scan_output_elts...). Each "
+		"scan_output is created by concatenating the value of the specified "
+		"scan_output_elt value at the end of each iteration of the loop. It is an error"
+		" if the dimensions of these values change across loop iterations.",
+		AttributeProto::GRAPH,
+		true)
+	.Attr(
+		"num_scan_inputs",
+		"An optional attribute specifying the number of scan_inputs M. "
+        "If neither this attribute nor the directions attribute is specified, the "
+        "number of scan_inputs M is assumed to be 1.",
+		AttributeProto::INT,
+        false
+	)
+    .Attr(
+		"directions",
+		"An optional list of M directions. The i-th element of the list specifies the direction "
+		"(forward or reverse) to be scanned for the i-th scan_input tensor. "
+        "If omitted, all scan_input tensors will be scanned in the forward direction.",
+		AttributeProto::STRINGS,
```

Comment:
I think it's better to use boolean or enum/int to indicate the direction rather than string.

Reply:
I agree … but my goal was to be consistent with the current RNN/LSTM/GRU spec. It might be better to change all of these uniformly, if we want to use an enum/int.

Reply:
Here there are only two choices, while for RNN there can be potentially many different activation functions.

Reply:
I was referring to the "direction: STRING" attribute (Specify if the RNN is forward, reverse, or bidirectional. Must be one of forward (default), reverse, or bidirectional.)

Reply:
But I can change it to Boolean … but the inconsistency might cause some confusion too.